Errata: Type additions (#1252)
php-fig/fig-standards: accepted/PSR-16-simple-cache-meta.md

* [Survey of existing cache implementations][1], by @dragoonis

[1]: https://docs.google.com/spreadsheet/ccc?key=0Ak2JdGialLildEM2UjlOdnA4ekg3R1Bfeng5eGlZc1E#gid=0

## 8. Errata
### 8.1 Type additions

> The 2.0 release of the `psr/simple-cache` package includs scalar parameter types and increases the minimum PHP version to 8.0 respectively.  This is considered a backwards compatible change for implementing libraries as PHP 7.2 introduces covariance for parameters.  Any implementation of 1.0 is compatible with 2.0. For calling libraries, however, this reduces the types that they may pass (as previously any parameter that could be cast to string could be accepted) and as such requires incrementing the major version.

Reviewer: ```suggestion
The 2.0 release of the `psr/simple-cache` package includes scalar parameter types and increases the minimum PHP version to 8.0.  This is considered a backwards compatible change for implementing libraries as PHP 7.2 introduces covariance for parameters.  Any implementation of 1.0 is compatible with 2.0. For calling libraries, however, this reduces the types that they may pass (as previously any parameter that could be cast to string could be accepted) and as such requires incrementing the major version.
```

Some typos.

> The 3.0 release includes return types.  Return types break backwards compatibility for implementing libraries as PHP does not support return type covariance.

Reviewer: This is not true.  PHP supports return type contravariance (or whatever the opposite of what parameters do is. I can never remember which is which.)  A child can add a more restrictive return type than its parent, but cannot widen it.

Author: ```suggestion
The 3.0 release includes return types.  Return types break backwards compatibility for implementing libraries as PHP does not support return type widening.
```

Implementing libraries **MAY** add return types to their own packages at their discretion, provided that:

* the return types match those in the 3.0 package.
* the implementation specifies a minimum PHP version of 8.0.0 or later
* the implementation depends on `"psr/simple-cache": "^2 || ^3"` so as to exclude the untyped 1.0 version.

Implementing libraries **MAY** add parameter types to their own package in a new minor release, either at the same time as adding return types or in a subsequent release, provided that:

* the parameter types match or widen those in the 2.0 package
* the implementation specifies a minimum PHP version of 8.0 if using mixed or union types or later.
* the implementation depends on `"psr/simple-cache": "^2 || ^3"` so as to exclude the untyped 1.0 version.

Implementing libraries are encouraged, but not required to transition their packages toward the 3.0 version of the package at their earliest convenience.

> Calling libraries are encouraged to ensure they are sending the correct types and to update their requirement to `"psr/simple-cache": "^1 || ^2 || ^3"` at their earliest convenience.

Reviewer: It's not actually possible to be compatible with version 1 and 3 at the same time.  That's the reason for the intermediary v2 step.

I think the instructions here are incorrect.  Please see the language in the other recently updated specs.

Author: A calling library can be compatible with versions 1 and 3, an implementing library can't.

Reviewer 2: Maybe we should underline that concept at the first mention of `calling library` then, since the difference can be summarized as "uses but don't implement the interfaces".

Author: @Jean85 Calling Library is defined [in the standard under "Definitions"](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-16-simple-cache.md)

* [Survey of existing cache implementations][1], by @dragoonis

[1]: https://docs.google.com/spreadsheet/ccc?key=0Ak2JdGialLildEM2UjlOdnA4ekg3R1Bfeng5eGlZc1E#gid=0

## 8. Errata

> ### 8.1 Throwable

Reviewer: ```suggestion

### 8.1 Throwable
```

Minor style nit.
